core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -309,6 +321,104 @@ private void balance() {
             balanceByWeights();
         }
 
+        /**
+         * Makes a decision about moving a single shard to a different node to form a more
+         * optimally balanced cluster.  This method is invoked from the cluster allocation
+         * explain API only.
+         */
+        private RebalanceDecision decideRebalance(final ShardRouting shard) {
+            if (shard.started() == false) {
+                // cannot rebalance a shard that isn't started
+                return RebalanceDecision.NOT_TAKEN;
+            }
+
+            Decision canRebalance = allocation.deciders().canRebalance(shard, allocation);
+            if (canRebalance.type() != Type.YES) {
+                // pick the first NO rebalance decision and use its explanation for the final explanation
+                String explanation = null;
+                Type explanationCause = null;
+                for (Decision subDecision : canRebalance.getDecisions()) {
+                    if ((subDecision.type() == Type.NO && (explanation == null || explanationCause == Type.THROTTLE))
+                            || (subDecision.type() == Type.THROTTLE && explanation == null)) {
+                        explanation = subDecision.getExplanation();
+                        explanationCause = subDecision.type();
+                    }
+                }
+                return new RebalanceDecision(canRebalance, Type.NO, explanation);
+            }
+
+            if (allocation.hasPendingAsyncFetch()) {
```

Comment:
I wanted the `canRebalance` decision go into the explanation even if there is a pending async shard fetch, so that's why I put this check after the `canRebalance` check.  But if you think that will confuse a user reading the output, then I can change it so the pending async fetch doesn't contain the `canRebalance` decision